pkg/util/tls/tls_integration_test.go

```
@@ -0,0 +1,484 @@
+	require.NoError(t, err)
+
+	cfg.HTTPListenPort = httpPort
+	cfg.GRPCListenPort = grpcPort
+
+	serv, err := server.New(cfg)
+	require.NoError(t, err)
+
+	serv.HTTP.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "OK")
+	})
+
+	grpc_health_v1.RegisterHealthServer(serv.GRPC, &grpcHealthCheck{true})
+
+	go func() {
+		err := serv.Run()
+		require.NoError(t, err)
+	}()
+
+	httpURL := fmt.Sprintf("https://localhost:%d/hello", httpPort)
+	grpcHost := fmt.Sprintf("localhost:%d", grpcPort)
+
+	for _, tc := range tcs {
+		tlsClientConfig, err := tc.clientConfig.GetTLSConfig()
+		require.NoError(t, err)
+
+		// HTTP
+		t.Run("HTTP/"+tc.name, func(t *testing.T) {
+			transport := &http.Transport{TLSClientConfig: tlsClientConfig}
+			client := &http.Client{Transport: transport}
+
+			resp, err := client.Get(httpURL)
+			if err == nil {
+				defer resp.Body.Close()
+			}
+			if tc.httpExpectError != nil {
+				tc.httpExpectError(t, err)
+				return
+			}
+			if err != nil {
+				assert.NoError(t, err, tc.name)
+				return
+			}
+			body, err := ioutil.ReadAll(resp.Body)
+			assert.NoError(t, err, tc.name)
+
+			assert.Equal(t, []byte("OK"), body, tc.name)
+
+		})
+
+		// GRPC
+		t.Run("GRPC/"+tc.name, func(t *testing.T) {
+			clientConfig := grpcclient.Config{}
+			clientConfig.RegisterFlags(flag.NewFlagSet("fake", flag.ContinueOnError))
+
+			dialOptions, err := tc.clientConfig.GetGRPCDialOptions()
+			assert.NoError(t, err, tc.name)
+			dialOptions = append([]grpc.DialOption{grpc.WithDefaultCallOptions(clientConfig.CallOptions()...)}, dialOptions...)
+
+			conn, err := grpc.Dial(grpcHost, dialOptions...)
+			assert.NoError(t, err, tc.name)
+			require.NoError(t, err, tc.name)
+			require.NoError(t, err, tc.name)
+
+			client := grpc_health_v1.NewHealthClient(conn)
+
+			// TODO: Investigate why the client doesn't really receive the
+			// error about the bad certificate from the server side and just
+			// see connection closed instead
```

Anyone having an idea how we can propagate the server error of "bad certificate" in the GRPC case?

---

https://github.com/grpc/grpc-go/issues/622#issuecomment-233107191 suggests that you need to use a WithBlock() call option to get the certificate error. This might not be possible with the existing clients.

---

I will have to investigate a bit more but I have found that issue and tried `WithBlock()` and had the request being blocked for a while rather than having any change of error behaviour. I will take another look tomorrow